starknet_devnet/blueprints/rpc/schema.py

```diff
@@ -0,0 +1,228 @@
+
+    RPC spec is currently formatted in the way, that every `$ref` is made with respect to
+    #/components/schemas/SCHEMA_NAME
+    The dict with schemas has to follow the same structure using
+    nested dictionaries.
+    """
+
+    methods, schemas = _load_schemas()
+    base_schema = methods[name]["result"]["schema"]
+    base_schema["unevaluatedProperties"] = False
+
+    return {**base_schema, "components": {"schemas": schemas}}
+
+
+def _request_schemas_for_method(name: str) -> OrderedDictType[str, Any]:
+    """
+    Return a dict with correct structure for jsonschema validation.
+
+    {
+        "param1": { // schema 1 },
+        "param2: { // schema 2 },
+        // ...
+        "components": {
+            "schemas": {
+                // rest of the schemas
+                // ...
+            }
+        }
+    }
+
+    See _response_schema_for_method docstring for more detailed explanation.
+    """
+    methods, schemas = _load_schemas()
+    params_json: List[Dict[str, Any]] = methods[name]["params"]
+
+    request_schemas = OrderedDict()
+    for param in params_json:
+        name = param["name"]
+        is_required = param.get("required", False)
+        schema = {
+            **param["schema"],
+            "is_required": is_required,
+            "components": {"schemas": schemas},
+        }
+        request_schemas[name] = schema
+
+    return request_schemas
+
+
+def _construct_method_name(method_name: str) -> str:
+    if "starknet_" not in method_name:
+        method_name = "starknet_" + method_name
+    return method_name
+
+
+def _assert_valid_rpc_schema(data: Dict[str, Any], method_name: str):
+    """
+    Check if rpc response is valid against the schema for given method name
+    """
+    schema = _response_schema_for_method(_construct_method_name(method_name))
+    validate(data, schema=schema)
+
+
+def _assert_valid_rpc_request(*args, method_name: str, **kwargs):
+    """
+    Validate if RPC request (parameters) is correct.
+
+    Raise ValidationError if not.
+    """
+    schemas = _request_schemas_for_method(_construct_method_name(method_name))
+
+    if args and kwargs:
+        raise ValueError("Cannot validate schemas with both args and kwargs provided.")
+
+    if args:
+        if len(args) > len(schemas):
+            raise ValidationError("Too many arguments provided.")
+
+        for name, arg in zip_longest(schemas.keys(), args, fillvalue="missing"):
+            if arg == "missing":
+                raise ValidationError(f"Missing arg for {name}.")
+
+            validate(arg, schemas[name])
+        return
+
+    if kwargs:
+        if len(kwargs) > len(schemas):
+            raise ValidationError("Too many arguments provided.")
+
+        for name, schema in schemas.items():
+            if name not in kwargs:
+                if schema["is_required"]:
+                    raise ValidationError(f"Missing kwarg for {name}.")
```

Comment:
This is untested: there are no methods with optional parameters in the rpc currently.

Reply:
How about creating a `fake` RPC schema to test it? 😏

Reply:
Hmm, this could be done with some patching I think. I'll test that.